File: client/src/main/scala/stasis/client/staging/DefaultFileStaging.scala
package stasis.client.staging

import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption

import scala.concurrent.ExecutionContext
import scala.concurrent.Future

import org.apache.pekko.Done

import stasis.client.analysis.PlatformMetadata

class DefaultFileStaging(
  storeDirectory: Option[Path],
  prefix: String,
  suffix: String
)(implicit ec: ExecutionContext)
    extends FileStaging {
  override def temporary(): Future[Path] =
    Future {
      storeDirectory match {
        case Some(dir) => Files.createTempFile(dir, prefix, suffix, temporaryFileAttributes: _*)
        case None      => Files.createTempFile(prefix, suffix, temporaryFileAttributes: _*)
      }
    }

  override def discard(file: Path): Future[Done] =
    Future {
      val _ = Files.deleteIfExists(file)
      Done
    }

  override def destage(from: Path, to: Path): Future[Done] =
    Future {
      val _ = Files.move(from, to, StandardCopyOption.REPLACE_EXISTING)
      Done
    }

  private val temporaryFileAttributes =
    PlatformMetadata.current.ownerOnlyFileAttributes
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/staging/DefaultFileStaging.kt
package stasis.client_android.lib.staging

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.attribute.PosixFilePermissions

class DefaultFileStaging(
    private val storeDirectory: Path?,
    private val prefix: String,
    private val suffix: String
) : FileStaging {
    override suspend fun temporary(): Path = withContext(Dispatchers.IO) {
        when (storeDirectory) {
            null -> Files.createTempFile(prefix, suffix, temporaryFileAttributes)
            else -> Files.createTempFile(storeDirectory, prefix, suffix, temporaryFileAttributes)
        }
    }

    override suspend fun discard(file: Path): Unit = withContext(Dispatchers.IO) {
        Files.deleteIfExists(file)
    }

    override suspend fun destage(from: Path, to: Path): Unit = withContext(Dispatchers.IO) {
        Files.move(from, to, StandardCopyOption.REPLACE_EXISTING)
    }

    private val temporaryFilePermissions = "rw-------"

    private val temporaryFileAttributes =
        PosixFilePermissions.asFileAttribute(
            PosixFilePermissions.fromString(
                temporaryFilePermissions
            )
        )
}
